File: src/main/java/Button.java

public class Button {
    public interface ClickListener {
        void perform();
    }

    private ClickListener mClickListener = null;
    public void setClickListener(ClickListener clickListener) {
        mClickListener = clickListener;
    }

    public void click() {
        if (mClickListener != null)
            mClickListener.perform();
    }

    public static void main(String[] args) {
        Button button = new Button();
        button.setClickListener(() -> {
            System.out.println("Click");
        });

        button.click();
    }
}

File: src/main/kotlin/Button.kt
class Button {
    private var mOnClick: (() -> Unit)? = null
    fun setClickListener(onClick: () -> Unit) {
        mOnClick = onClick
    }

    fun click() {
        mOnClick?.invoke()
    }

}

fun main(args: Array<String>) {
    val button = Button()
    button.setClickListener { println("Click") }
    button.click()
}